Unable to access context.test in before_scenario

We've set up `before_scenario` to swap out Django's test client with Django Rest Framework's `APIClient`:

```
from rest_framework.test import APIClient

def before_scenario(context, scenario):
    context.test.client = APIClient()
```

This worked in 0.5.0. In 1.0.0, it seems that I can't access `context.test` at this point:

```
Exception AttributeError: 'PatchedContext' object has no attribute 'test'
Traceback (most recent call last):
  File "manage.py", line 30, in <module>
    main()
  File "manage.py", line 26, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/behave_django/management/commands/behave.py", line 128, in handle
    exit_status = behave_main(args=behave_args)
  File "/usr/local/lib/python3.6/site-packages/behave/__main__.py", line 109, in main
    failed = runner.run()
  File "/usr/local/lib/python3.6/site-packages/behave/runner.py", line 672, in run
    return self.run_with_paths()
  File "/usr/local/lib/python3.6/site-packages/behave/runner.py", line 693, in run_with_paths
    return self.run_model()
  File "/usr/local/lib/python3.6/site-packages/behave/runner.py", line 471, in run_model
    self.run_hook('before_all', context)
  File "/usr/local/lib/python3.6/site-packages/behave_django/environment.py", line 91, in run_hook
    behave_run_hook(self, name, context, *args)
  File "/usr/local/lib/python3.6/site-packages/behave/runner.py", line 405, in run_hook
    self.hooks[name](context, *args)
  File "features/environment.py", line 14, in before_all
    context.test.client = APIClient()
  File "/usr/local/lib/python3.6/site-packages/behave/runner.py", line 214, in __getattr__
    raise AttributeError(msg)
AttributeError: 'PatchedContext' object has no attribute 'test'
```

As far as I can tell, I should be able to access `context.test` here.

Now I get this error even without the -S flag:

```
Exception AssertionError: Web browser automation is not available. This scenario step can not be run with the --simple or -S flag.
Traceback (most recent call last):
  File "manage.py", line 28, in <module>
    execute_from_command_line(sys.argv)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\django\core\management\__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\django\core\management\__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\django\core\management\base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\django\core\management\base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave_django\management\commands\behave.py", line 128, in handle
    exit_status = behave_main(args=behave_args)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\__main__.py", line 109, in main
    failed = runner.run()
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\runner.py", line 672, in run
    return self.run_with_paths()
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\runner.py", line 693, in run_with_paths
    return self.run_model()
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\runner.py", line 483, in run_model
    failed = feature.run(self)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\model.py", line 523, in run
    failed = scenario.run(runner)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\model.py", line 867, in run
    runner.run_hook('before_scenario', runner.context, self)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave_django\environment.py", line 104, in run_hook
    behave_run_hook(self, name, context, *args)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave\runner.py", line 405, in run_hook
    self.hooks[name](context, *args)
  File "entry\features\environment.py", line 40, in before_scenario
    context.driver.get(context.base_url)
  File "C:\Users\Datacenter\AppData\Local\Programs\Python\Python35\lib\site-packages\behave_django\environment.py", line 11, in base_url
    'Web browser automation is not available.'
AssertionError: Web browser automation is not available. This scenario step can not be run with the --simple or -S flag.
```
The test is now accessible but there is no live_server_url which causes the error.

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -1,6 +1,14 @@
 Release History
 ---------------
 
+Unreleased
+++++++++++
+
+**Features and Improvements**
+
+- Added :code:`django_ready` hook for running setup code within the django environment
+
+
 1.0.0 (2017-10-25)
 ++++++++++++++++++
 

--- a/behave_django/environment.py
+++ b/behave_django/environment.py
@@ -18,6 +18,18 @@ class PatchedContext(Context):
             resolve_url(to, *args, **kwargs) if to else '')
 
 
+def load_registered_fixtures(context):
+    """
+    Apply fixtures that are registered with the @fixtures decorator.
+    """
+    for step in context.scenario.all_steps:
+        match = step_registry.registry.find_match(step)
+        if match and hasattr(match.func, 'registered_fixtures'):
+            if not context.test.fixtures:
+                context.test.fixtures = []
+            context.test.fixtures.extend(match.func.registered_fixtures)
+
+
 class BehaveHooksMixin(object):
     """
     Provides methods that run during test execution
@@ -37,14 +49,16 @@ class BehaveHooksMixin(object):
         # because behave.runner.Context.__setattr__ is implemented wrongly.
         object.__setattr__(context, '__class__', PatchedContext)
 
-    def before_scenario(self, context):
+    def setup_testclass(self, context):
         """
-        Method that runs immediately before behave's before_scenario function
-
-        Sets up the test case.
+        Adds the test instance to context
         """
         context.test = self.testcase_class()
 
+    def setup_fixtures(self, context):
+        """
+        Sets up fixtures
+        """
         if getattr(context, 'fixtures', None):
             context.test.fixtures = context.fixtures
 
@@ -52,30 +66,26 @@ class BehaveHooksMixin(object):
             context.test.reset_sequences = context.reset_sequences
 
         if hasattr(context, 'scenario'):
-            self.load_registered_fixtures(context=context)
+            load_registered_fixtures(context)
 
+    def setup_test(self, context):
+        """
+        Sets up the Django test
+
+        This method runs the code necessary to create the test database, start
+        the live server, etc.
+        """
         context.test._pre_setup(run=True)
         context.test.setUpClass()
         context.test()
 
-    def after_scenario(self, context):
+    def teardown_test(self, context):
         """
-        Method that runs immediately after behave's after_scenario function
+        Tears down the Django test
         """
         context.test.tearDownClass()
         context.test._post_teardown(run=True)
         del context.test
-
-    def load_registered_fixtures(self, context):
-        """
-        Apply fixtures that are registered with the @fixtures decorator.
-        """
-        for step in context.scenario.all_steps:
-            match = step_registry.registry.find_match(step)
-            if match and hasattr(match.func, 'registered_fixtures'):
-                if not context.test.fixtures:
-                    context.test.fixtures = []
-                context.test.fixtures.extend(match.func.registered_fixtures)
 
 
 def monkey_patch_behave(django_test_runner):
@@ -91,9 +101,12 @@ def monkey_patch_behave(django_test_runner):
         behave_run_hook(self, name, context, *args)
 
         if name == 'before_scenario':
-            django_test_runner.before_scenario(context)
+            django_test_runner.setup_testclass(context)
+            django_test_runner.setup_fixtures(context)
+            django_test_runner.setup_test(context)
+            behave_run_hook(self, 'django_ready', context)
 
         if name == 'after_scenario':
-            django_test_runner.after_scenario(context)
+            django_test_runner.teardown_test(context)
 
     ModelRunner.run_hook = run_hook

--- a/docs/usage.rst
+++ b/docs/usage.rst
@@ -36,9 +36,8 @@ Examples:
 Django’s testing client
 -----------------------
 
-Attached to the context is an instance of TestCase.  You can access it
-via ``context.test``.  This means you can do things like use Django’s
-testing client.
+Internally, Django's TestCase is used to maintain the test environment. You can
+access the TestCase instance via ``context.test``.
 
 .. code-block:: python
 
@@ -47,7 +46,6 @@ testing client.
     def visit(context, url):
         # save response in context for next step
         context.response = context.test.client.get(url)
-
 
 Simple testing
 --------------
@@ -90,18 +88,35 @@ regular TestCases.  So you can do something like:
 
 And you don’t have to clean the database yourself.
 
-If you have `factories`_ you want to instantiate on a per-scenario basis,
-you can initialize them in ``environment.py`` like this:
+django_ready hook
+-----------------
+
+You can add a ``django_ready`` function in your ``environment.py`` file in case
+you want to make per-scenario changes inside a transaction.
+
+For example, if you have `factories`_ you want to instantiate on a per-scenario
+basis, you can initialize them in ``environment.py`` like this:
 
 .. code-block:: python
 
     from myapp.main.tests.factories import UserFactory, RandomContentFactory
 
 
-    def before_scenario(context, scenario):
+    def django_ready(context, scenario):
+        # This function is run inside the transaction
         UserFactory(username='user1')
         UserFactory(username='user2')
         RandomContentFactory()
+
+Or maybe you want to modify the ``test`` instance:
+
+.. code-block:: python
+
+    from rest_framework.test import APIClient
+
+
+    def django_ready(context, scenario):
+        context.test.client = APIClient()
 
 Fixture loading
 ---------------
@@ -110,7 +125,7 @@ behave-django can load your fixtures for you per feature/scenario. There are
 two approaches to this:
 
 * loading the fixtures in ``environment.py``, or
-* using a decorator on your step method
+* using a decorator on your step function
 
 
 Fixtures in environment.py
@@ -145,8 +160,8 @@ You could also have fixtures per Feature too
             # This works because behave will use the same context for
             # everything below Feature. (Scenarios, Outlines, Backgrounds)
 
-Of course, since ``context.fixtures`` is really just a list, you can
-mutate it however you want, it will only be processed upon leaving the
+Of course, since ``context.fixtures`` is really just a list, you can mutate it
+however you want, it will only be processed upon leaving the
 ``before_scenario()`` function of your ``environment.py`` file.
 
 .. note::
@@ -160,7 +175,7 @@ mutate it however you want, it will only be processed upon leaving the
 Fixtures using a decorator
 **************************
 
-You can define `Django fixtures`_ using a method decorator. The decorator will
+You can define `Django fixtures`_ using a function decorator. The decorator will
 load the fixtures in the ``before_scenario``, as documented above. It is merely
 a convenient way to keep fixtures close to your steps.
 

--- a/features/environment.py
+++ b/features/environment.py
@@ -15,3 +15,7 @@ def before_scenario(context, scenario):
     if scenario.name == 'Load fixtures then reset sequences':
         context.fixtures.append('behave-second-fixture.json')
         context.reset_sequences = True
+
+
+def django_ready(context):
+    context.django = True

--- a/features/running-tests.feature
+++ b/features/running-tests.feature
@@ -7,3 +7,7 @@ Feature: Running tests
         Given this step exists
         When I run "python manage.py behave"
         Then I should see the behave tests run
+
+    Scenario: Test django_ready
+        When I run "python manage.py behave"
+        Then django_ready should be called

--- a/features/steps/running_tests.py
+++ b/features/steps/running_tests.py
@@ -14,3 +14,8 @@ def run_command(context):
 @then(u'I should see the behave tests run')
 def is_running(context):
     pass
+
+
+@then(u'django_ready should be called')
+def django_context(context):
+    assert context.django

--- a/tests/test_simple_testcase.py
+++ b/tests/test_simple_testcase.py
@@ -24,14 +24,14 @@ class TestSimpleTestCase(DjangoSetupMixin):
     def test_simple_test_runner_uses_simple_testcase(self):
         runner = mock.MagicMock()
         context = Context(runner)
-        SimpleTestRunner().before_scenario(context)
+        SimpleTestRunner().setup_testclass(context)
         assert isinstance(context.test, TestCase)
 
     def test_simple_testcase_fails_when_accessing_base_url(self):
         runner = Runner(mock.MagicMock())
         runner.context = Context(runner)
         SimpleTestRunner().patch_context(runner.context)
-        SimpleTestRunner().before_scenario(runner.context)
+        SimpleTestRunner().setup_testclass(runner.context)
         with pytest.raises(AssertionError):
             assert runner.context.base_url == 'should raise an exception!'
 
@@ -39,6 +39,6 @@ class TestSimpleTestCase(DjangoSetupMixin):
         runner = Runner(mock.MagicMock())
         runner.context = Context(runner)
         SimpleTestRunner().patch_context(runner.context)
-        SimpleTestRunner().before_scenario(runner.context)
+        SimpleTestRunner().setup_testclass(runner.context)
         with pytest.raises(AssertionError):
             runner.context.get_url()

